src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -384,59 +384,69 @@ def process_messages(
 
     org_strings = defaultdict(set)
     strings = set()
-    skipped_offsets = set()
+    skipped_offsets: Mapping[int, Set[int]] = defaultdict(set)
     with metrics.timer("process_messages.parse_outer_message"):
-        parsed_payloads_by_offset: MutableMapping[int, json.JSONData] = {}
+        parsed_payloads_by_offset: MutableMapping[
+            int, MutableMapping[int, json.JSONData]
+        ] = defaultdict(dict)
+
         for msg in outer_message.payload:
             try:
                 parsed_payload = json.loads(msg.payload.value.decode("utf-8"), use_rapid_json=True)
-                parsed_payloads_by_offset[msg.offset] = parsed_payload
+                parsed_payloads_by_offset[msg.partition.index][msg.offset] = parsed_payload
             except rapidjson.JSONDecodeError:
-                skipped_offsets.add(msg.offset)
+                skipped_offsets[msg.partition.index].add(msg.offset)
                 logger.error(
                     "process_messages.invalid_json",
                     extra={"payload_value": str(msg.payload.value)},
                     exc_info=True,
                 )
                 continue
 
-        for offset, message in parsed_payloads_by_offset.items():
-            metric_name = message["name"]
-            org_id = message["org_id"]
-            tags = message.get("tags", {})
-
-            if not valid_metric_name(metric_name):
-                logger.error(
-                    "process_messages.invalid_metric_name",
-                    extra={"org_id": org_id, "metric_name": metric_name, "offset": offset},
-                )
-                skipped_offsets.add(offset)
-                continue
-
-            invalid_strs = invalid_metric_tags(tags)
-
-            if invalid_strs:
-                # sentry doesn't seem to actually capture nested logger.error extra args
-                sentry_sdk.set_extra("all_metric_tags", tags)
-                logger.error(
-                    "process_messages.invalid_tags",
-                    extra={
-                        "org_id": org_id,
-                        "metric_name": metric_name,
-                        "invalid_tags": invalid_strs,
-                        "offset": offset,
-                    },
-                )
-                skipped_offsets.add(offset)
-                continue
-
-            parsed_strings = {
-                metric_name,
-                *tags.keys(),
-                *tags.values(),
-            }
-            org_strings[org_id].update(parsed_strings)
-            strings.update(parsed_strings)
+        for partition_idx in parsed_payloads_by_offset:
+            for offset, message in parsed_payloads_by_offset[partition_idx].items():
```

Comment:
I don't really like this nested for loop, but the number of partitions will be relatively low